firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,186 @@
+      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in its payload is independent of the data in input_buffer")
+
+    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports invalid crc status")
+      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
+      THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload given to it's constructor is empty")
+      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in its payload is independent of the data in input_buffer")
+
+  GIVEN("A CRC element receiver of capacity 254 bytes and output_crcelement constructed with an non-empty payload buffer")
+
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is given to the receiver")
+      THEN("the transform method reports ok status")
+      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
+      THEN("the value returned by it's crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload given to it's constructor is '0x01 0x05 0x12 0x13 0x14 0x15 0x16'")
+      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in its payload is independent of the data in input_buffer")
+  
+    WHEN("An empty input_buffer is given to the crc element receiver")
+      THEN("the transform method reports invalid parse status")
+      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
+      THEN("it's crc accessor method returns 0")
+      THEN("the payload given to it's constructor is '0x01 0x05 0x12 0x13 0x14 0x15 0x16'")
+      THEN("Its payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in its payload is independent of the data in input_buffer")
+
+    WHEN("An empty input_buffer is given to the crc element receiver and output_crcelement with a non-zero crc member")
+      THEN("The write method of output_crcelement reports ok status")
+      THEN("The value returned by the crc accessor method of output_crcelement is '0x40D06D79'")
+      THEN("the transform status is equal to invalid parse")
+      THEN("After the transform method is called, The ParsedCRCElement output parameter has the following properties : ")
+      THEN("the value returned by it's crc accessor method is equal to '0x40D06D79'")
+      THEN("the payload given to it's constructor is '0x01 0x05 0x12 0x13 0x14 0x15 0x16'")
```

Comment:
Are these values coming from the input_buffer after a successful parse, or are they the data which was already in output_crcelement and which was left unchanged after an unsuccessful parse? The CRC value appears to be the same before and after transform, while the payload is left unspecified before transform.

I think it'll be clearer and more maintainable if you use the following structure for THENs in this file (don't copy these literally, adapt the language and assertions to make sense for each specific test):
- THEN the value returned by output_crcelement's crc accessor method remains the same as before the transform method call
- THEN the value returned by output_crcelement's crc accessor method becomes equal to the crc field of input_buffer's header
- THEN the value returned by output_crcelement's payload accessor method remains the same as before the transform method call
- THEN the value returned by output_crcelement's payload accessor method becomes equal to the payload of input_buffer